bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py

```diff
@@ -0,0 +1,512 @@
+            self.node_id += 1
+        else:
+            # Action or instant action
+
+            # The result is stored in the vector of results to be returned
+            self.graph_nodes += '\n  result_handles.emplace_back(res_ptr{});' \
+                                .format(self.res_ptr_id)
+
+            # The result type is stored in the vector of result types to be
+            # returned
+            self.graph_nodes += \
+                '\n  auto c{} = TClass::GetClass(typeid(res_ptr{}));' \
+                .format(self.res_ptr_id, self.res_ptr_id)
+            self.graph_nodes += \
+                '\n  result_types.emplace_back(c{}->GetName());' \
+                .format(self.res_ptr_id)
+
+            self.res_ptr_id += 1
+            return None
+
+        return new_node_id
+
+    def _handle_snapshot(self, operation, range_id):
+        '''
+        Does three extra settings needed for Snapshot nodes:
+        - Modifies the output file name to be of the form `filename_rangeid`,
+        since it is a partial snapshot for a given range.
+        - Stores the index of the returned vector<RResultHandle> in which the
+        result of this Snapshot is stored, together with the modified file
+        path.
+
+        Args:
+            operation (Operation): object representing the operation to be added
+                to the graph.
+
+            range_id (int): id of the current range. Needed to assign a name to
+                a partial Snapshot output file.
+        '''
+
+        # Modify file name
+        filename = operation.args[1].partition('.root')[0]
+        path_with_range = '{filename}_{rangeid}.root' \
+                          .format(filename=filename, rangeid=range_id)
+        operation.args[1] = path_with_range
+
+        # Store Snapshot result index -> path
+        self.snapshots.append(
+            CppWorkflow.SnapshotData(self.res_ptr_id, path_with_range))
+
+    def _handle_asnumpy(self, operation, parent_id):
+        '''
+        Since AsNumpy is a Python-only action, it can't be included in the
+        C++ workflow built by this class. Therefore, this function takes care
+        of saving the RDF node, generated in C++, on which an AsNumpy action
+        should be applied from Python.
+        This function also stores the index of the AsNumpy result in the final
+        list of results, where the index depends on the DFS traversal of the
+        computation graph.
+
+        Args:
+            operation (Operation): object representing the AsNumpy operation
+            
+            parent_id (int): id of the parent node in this workflow. Used to
+                save that node in the vector of output nodes.
+        '''
+
+        self.py_actions.append(CppWorkflow.PyActionData(self.res_ptr_id, operation))
+        self.graph_nodes += \
+            '\n  output_nodes.push_back(ROOT::RDF::AsRNode(rdf{}));' \
+            .format(parent_id)
+        self.res_ptr_id += 1
```

Comment:
if I understand correctly, we increment `res_ptr_id` here because we know that _eventually_ we will insert a result at position `i` for this action. Until then, indexing into the list of results using `res_ptr_id` might be tricky, e.g. we have to make sure to add the result for this action in the list before we do a substitution of the results of Snapshot.

This implicit required sequencing seems easy to forget and break when refactoring, could we instead already add dummy contents to the relevant vectors so we later we only have to substitute contents and it doesn't matter in which order we do it?

Reply:
Yes your reasoning is right, the insertion of the AsNumpy results needs to happen before the Snapshots are treated. And yes indeed I thought of the problem you just described, but I'd need to create an empty `RResultHandle` constructor just for that (the AsNumpy dummies). If you think that's preferable, I can change it.

Reply:
you don't _need_ an empty RResultHandle constructor, you can construct a RResultHandle from an empty RResultPtr instead -- anyway why not: https://github.com/root-project/root/pull/8979

Reply:
I think it's better this way, otherwise we'd be creating a dummy inside a dummy.